R/PipeOpLearnerCVPlus.R

```diff
@@ -0,0 +1,204 @@
+#' library("mlr3")
+#'
+#' task = tsk("mtcars")
+#' learner = lrn("regr.ranger")
+#' lrncvplus_po = mlr_pipeops$get("learner_cv_plus", learner)
+#'
+#' lrncvplus_po$train(list(task))
+#' lrncvplus_po$predict(list(task))
+#' \dontshow{ \} }
+PipeOpLearnerCVPlus = R6Class("PipeOpLearnerCVPlus",
+  inherit = PipeOp,
+  public = list(
+    initialize = function(learner, id = NULL, param_vals = list()) {
+      private$.learner = as_learner(learner, clone = TRUE)
+      id = id %??% private$.learner$id
+      type = private$.learner$task_type
+
+      if (!("regr" %in% type)) {
+        stop("PipeOpLearnerCVPlus only supports regression.")
+      }
+
+      task_type = mlr_reflections$task_types[type, mult = "first"]$task
+      out_type = mlr_reflections$task_types[type, mult = "first"]$task
+
+      # paradox requirements 1.0
+      private$.cvplus_param_set = ps(
+        folds = p_int(lower = 2L, upper = Inf, tags = c("train", "required")),
+        alpha = p_dbl(lower = 0L, upper = 1L, tags = c("predict", "required"))
+      )
+
+      private$.cvplus_param_set$values = list(folds = 3, alpha = 0.05) # default
+
+      super$initialize(id, param_set = alist(cvplus = private$.cvplus_param_set, private$.learner$param_set),
+                        param_vals = param_vals,
+                        input = data.table(name = "input", train = task_type, predict = task_type),
+                        output = data.table(name = "output", train = "NULL", predict = out_type),
+                        packages = learner$packages,
+                        tags = c("learner", "ensemble")
+      )
+    }
+  ),
+  active = list(
+    learner = function(val) {
+      if (!missing(val)) {
+        if (!identical(val, private$.learner)) {
+          stop("$learner is read-only.")
+        }
+      }
+      private$.learner
+    },
+    learner_model = function(val) {
+      if (!missing(val)) {
+        if (!identical(val, private$.learner)) {
+          stop("$learner_model is read-only.")
+        }
+      }
+      if (is.null(self$state) || is_noop(self$state)) {
+        private$.learner
+      } else {
+        multiplicity_recurse(self$state, clone_with_state, learner = private$.learner)
+      }
+    },
+    predict_type = function(val) {
+      if (!missing(val)) {
+        stop("$predict_type is read-only.")
+      }
+      mlr_reflections$learner_predict_types$regr$quantiles  # Returns c("response", "quantiles")
+    }
+  ),
+  private = list(
+    .state_class = "pipeop_learner_cv_plus_state",
+
+    .train = function(task) {
+      pv = private$.cvplus_param_set$values
+
+      # Compute CV Predictions
+      rdesc = rsmp("cv", folds = pv$folds)
+      rr = resample(task, private$.learner, rdesc, store_models = TRUE)
+
+      prds = rbindlist(map(rr$predictions(predict_sets = "test"), as.data.table), idcol = "fold")
+
+      # Add trained models and residuals to PipeOp state
+      self$state = list(cv_models = rr$learners,
```

Comment:
```suggestion
      self$state = list(cv_model_states = map(rr$learners, "state"),
```

Reply:
It is best to only store the essential information in the PipeOp's `$state`, which are the model states. Try not to store the entire learner objects.

Reply:
You'd also need to update the documentation.

Reply:
Okay, I changed the state and documentation to only include the state.